File: src-tauri/src/domain/report.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GeneratedReport {
    pub title: String,
    pub start_date: String,
    pub end_date: String,
    pub recipient_name: Option<String>,
    pub content: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerateReportInput {
    pub start_date: String,
    pub end_date: String,
    pub recipient_name: Option<String>,
    pub project_ids: Option<Vec<String>>,
    pub include_commits: Option<bool>,
    pub include_manual_logs: Option<bool>,
    pub include_weekly_tasks: Option<bool>,
    pub include_hidden: Option<bool>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Report {
    pub id: String,
    pub title: String,
    pub start_date: String,
    pub end_date: String,
    pub recipient_name: Option<String>,
    pub content: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportSummary {
    pub id: String,
    pub title: String,
    pub start_date: String,
    pub end_date: String,
    pub recipient_name: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveReportInput {
    pub title: String,
    pub start_date: String,
    pub end_date: String,
    pub recipient_name: Option<String>,
    pub content: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportItem {
    pub id: String,
    pub report_id: String,
    pub project_id: Option<String>,
    pub source_type: String,
    pub source_id: Option<String>,
    pub summary: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateReportItemInput {
    pub report_id: String,
    pub project_id: Option<String>,
    pub source_type: String,
    pub source_id: Option<String>,
    pub summary: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportNote {
    pub id: String,
    pub project_id: Option<String>,
    pub note_type: String,
    pub date: String,
    pub content: String,
    pub included_in_report: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListReportNotesInput {
    pub from: String,
    pub to: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateReportNoteInput {
    pub project_id: Option<String>,
    pub note_type: String,
    pub date: String,
    pub content: String,
    pub included_in_report: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateReportNoteInput {
    pub project_id: Option<String>,
    pub note_type: Option<String>,
    pub date: Option<String>,
    pub content: Option<String>,
    pub included_in_report: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveDailyReviewNoteInput {
    pub date: String,
    pub finished: String,
    pub blocked: String,
    pub carry_into_tomorrow: String,
    pub included_in_report: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReportAiProvider {
    LocalLlamaCpp,
    OpenrouterFree,
    Groq,
}

impl ReportAiProvider {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::LocalLlamaCpp => "local_llama_cpp",
            Self::OpenrouterFree => "openrouter_free",
            Self::Groq => "groq",
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportPolishInput {
    pub draft: String,
    pub start_date: String,
    pub end_date: String,
    pub recipient_name: Option<String>,
    pub project_ids: Option<Vec<String>>,
    pub include_hidden: Option<bool>,
    pub provider: Option<ReportAiProvider>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportPolishResult {
    pub content: String,
    pub provider: String,
    pub model: String,
    pub used_fallback: bool,
    pub message: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportReadinessInput {
    pub start_date: String,
    pub end_date: String,
    pub project_ids: Option<Vec<String>>,
    pub include_hidden: Option<bool>,
    pub provider: Option<ReportAiProvider>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportReadinessFinding {
    pub severity: String,
    pub title: String,
    pub detail: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportReadinessAnalysis {
    pub provider: String,
    pub model: String,
    pub score: i32,
    pub summary: String,
    pub findings: Vec<ReportReadinessFinding>,
    pub used_fallback: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportAiProviderStatus {
    pub provider: String,
    pub available: bool,
    pub configured: bool,
    pub online: bool,
    pub model: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportAiStatus {
    pub enabled: bool,
    pub preferred_provider: String,
    pub providers: Vec<ReportAiProviderStatus>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportAiModel {
    pub id: String,
    pub name: String,
    pub provider: String,
    pub context_length: Option<i64>,
    pub description: Option<String>,
    pub input_price: Option<String>,
    pub output_price: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReportAiModelList {
    pub provider: String,
    pub models: Vec<ReportAiModel>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectReportAiProviderInput {
    pub provider: ReportAiProvider,
    pub api_key: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TestReportAiProviderInput {
    pub provider: ReportAiProvider,
}

File: src/types/report.ts
import type { GitRefFilter } from "./project";
import type { ProjectClassification } from "./project";

export type GeneratedReport = {
  title: string;
  startDate: string;
  endDate: string;
  recipientName?: string | null;
  content: string;
};

export type Report = GeneratedReport & {
  id: string;
  createdAt: string;
};

export type ReportSummary = {
  id: string;
  title: string;
  startDate: string;
  endDate: string;
  recipientName?: string | null;
  createdAt: string;
};

export type GenerateReportInput = {
  startDate: string;
  endDate: string;
  recipientName?: string | null;
  projectIds?: string[] | null;
  classification?: ProjectClassification | null;
  gitRefs?: GitRefFilter[] | null;
  worktreePaths?: string[] | null;
  useProjectGitFocus?: boolean | null;
  includeCommits?: boolean | null;
  includeManualLogs?: boolean | null;
  includeWeeklyTasks?: boolean | null;
  includeHidden?: boolean | null;
};

export type SaveReportInput = {
  title: string;
  startDate: string;
  endDate: string;
  recipientName?: string | null;
  content: string;
};

export type ReportNote = {
  id: string;
  projectId?: string | null;
  noteType: string;
  date: string;
  content: string;
  includedInReport: boolean;
  createdAt: string;
  updatedAt: string;
};

export type ListReportNotesInput = {
  from: string;
  to: string;
};

export type SaveDailyReviewNoteInput = {
  date: string;
  finished: string;
  blocked: string;
  carryIntoTomorrow: string;
  includedInReport?: boolean | null;
};

export type ReportAiProvider =
  | "local_llama_cpp"
  | "openrouter_free"
  | "groq"
  | "nvidia_build";

export type ReportPolishInput = {
  draft: string;
  startDate: string;
  endDate: string;
  recipientName?: string | null;
  projectIds?: string[] | null;
  classification?: ProjectClassification | null;
  gitRefs?: GitRefFilter[] | null;
  worktreePaths?: string[] | null;
  useProjectGitFocus?: boolean | null;
  includeHidden?: boolean | null;
  provider?: ReportAiProvider | null;
  streamId?: string | null;
};

export type ReportPolishResult = {
  content: string;
  provider: string;
  model: string;
  usedFallback: boolean;
  fallbackReason:
    | "none"
    | "provider_unavailable"
    | "provider_error"
    | "empty_output"
    | "length_limit"
    | "cancelled";
  message: string;
  diagnostics?: string | null;
};

export type ReportAiStreamPayload = {
  streamId: string;
  eventType: "start" | "delta" | "reasoning" | "done" | "error" | "cancelled";
  content: string;
  message?: string | null;
};

export type CancelReportAiStreamInput = {
  streamId: string;
};

export type ReportReadinessInput = {
  startDate: string;
  endDate: string;
  projectIds?: string[] | null;
  classification?: ProjectClassification | null;
  gitRefs?: GitRefFilter[] | null;
  worktreePaths?: string[] | null;
  useProjectGitFocus?: boolean | null;
  includeHidden?: boolean | null;
  provider?: ReportAiProvider | null;
};

export type ReportReadinessFinding = {
  severity: string;
  title: string;
  detail: string;
};

export type ReportReadinessAnalysis = {
  provider: string;
  model: string;
  score: number;
  summary: string;
  findings: ReportReadinessFinding[];
  usedFallback: boolean;
};

export type ReportAiProviderStatus = {
  provider: string;
  available: boolean;
  configured: boolean;
  online: boolean;
  model: string;
  message: string;
};

export type ReportAiStatus = {
  enabled: boolean;
  preferredProvider: string;
  providers: ReportAiProviderStatus[];
};

export type ReportAiModel = {
  id: string;
  name: string;
  provider: string;
  contextLength?: number | null;
  description?: string | null;
  inputPrice?: string | null;
  outputPrice?: string | null;
};

export type ReportAiModelList = {
  provider: string;
  models: ReportAiModel[];
};

export type ConnectReportAiProviderInput = {
  provider: ReportAiProvider;
  apiKey: string;
};

export type TestReportAiProviderInput = {
  provider: ReportAiProvider;
};
